Add Labels to Frontend

miles-grant-ibigroup opened the pull request:
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JSDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This makes use of https://github.com/ibi-group/datatools-server/pull/403 to add label functionality for feed sources. Feed sources can have labels applied to them, feed sources can be filtered by label. Labels can be managed in the dashboard. New components have been created to add a new Label, and to assign labels to a feed source.

#### Notes
When assigning labels from the feed sources dashboard, the assigner disappears after each click. This is due to how the existing redux actions are set up. It's possible to create a new action that updates things differently, but I don't want to waste time on that if it isn't needed.

evansiroky commented on lib/common/components/FeedLabel.js:
```diff
@@ -0,0 +1,104 @@
+            />
+            <LabelEditorModal ref='editModal' label={this.props.label} />
+
+            <button onClick={() => this._onClickEdit()}>
```
The `_onClickEdit` method is already bound to the instance since it is defined in the style `_onClickEdit = () =>`, so this could be:

```suggestion
            <button onClick={this._onClickEdit}>
```

miles-grant-ibigroup replied:
Good catch thank you! Fixed in a1bc3e9

evansiroky commented on lib/common/components/FeedLabel.js:
```diff
@@ -0,0 +1,104 @@
+            <button onClick={() => this._onClickEdit()}>
+              <Icon type='pencil' />
+            </button>
+            <button onClick={() => this._onClickDelete()}>
```
The `_onClickDelete ` method is already bound to the instance since it is defined in the style `_onClickDelete = () =>`, so this could be:
```suggestion
            <button onClick={this._onClickDelete}>
```

miles-grant-ibigroup replied:
Fixed in a1bc3e9

evansiroky commented on lib/common/containers/FeedLabel.js:
```diff
@@ -0,0 +1,21 @@
+export default connect(
+  mapStateToProps,
+  mapDispatchToProps
+)(FeedLabel)
```
In practice with newer components, I believe we've abandoned the separation of components into connected containers and the components, so it's fine to combine this in just the component file.

miles-grant-ibigroup replied:
Ah thank you! Refactored in a1bc3e9

evansiroky commented on lib/manager/containers/LabelAssigner.js:
```diff
@@ -0,0 +1,20 @@
+export default connect(
+  mapStateToProps,
+  mapDispatchToProps
+)(LabelAssigner)
```
In practice with newer components, I believe we've abandoned the separation of components into connected containers and the components, so it's fine to combine this in just the component file.

miles-grant-ibigroup replied:
Refactored in a1bc3e9

evansiroky commented on lib/manager/containers/LabelEditor.js:
```diff
@@ -0,0 +1,22 @@
+export default connect(
+  mapStateToProps,
+  mapDispatchToProps
+)(LabelEditor)
```
In practice with newer components, I believe we've abandoned the separation of components into connected containers and the components, so it's fine to combine this in just the component file.

miles-grant-ibigroup replied:
Refactored in a1bc3e9

evansiroky commented on yarn.lock:
```diff
@@ -14008,6 +14008,11 @@ tinycolor2@^1.4.1:
   resolved "https://registry.yarnpkg.com/tinycolor2/-/tinycolor2-1.4.1.tgz#f4fad333447bc0b07d4dc8e9209d8f39a8ac77e8"
   integrity sha1-9PrTM0R7wLB9TcjpIJ2POaisd+g=
 
+tinycolor2@^1.4.2:
```
Can you deduplicate the tinycolor resolutions so that just the most recent one is used?

miles-grant-ibigroup replied:
Done in a1bc3e9

evansiroky commented on lib/manager/actions/feeds.js:
```diff
@@ -81,6 +82,10 @@ export function createFeedSource (newFeed: NewFeed) {
 export function updateFeedSource (feedSource: Feed, properties: {[string]: any}) {
   return function (dispatch: dispatchFn, getState: getStateFn) {
     dispatch(savingFeedSource())
+    // Convert all labels to be only strings, which the server requires
+    feedSource.labels = stringifyLabels(feedSource.labels)
+    properties.labels = stringifyLabels(properties.labels)
```
Please remove if refactoring data model as recommended in https://github.com/ibi-group/datatools-server/pull/403#pullrequestreview-714190079.

evansiroky commented on lib/types/index.js:
```diff
@@ -385,6 +384,7 @@ export type Feed = {
   id: string,
   isCreating?: boolean,
   isPublic: boolean,
+  labels: Array<Label>,
```
Please remove if refactoring data model as recommended in https://github.com/ibi-group/datatools-server/pull/403#pullrequestreview-714190079.

evansiroky commented on lib/style.css:
```diff
@@ -141,6 +143,114 @@ td.feed-source-info {
 
+/* LABELS */
+.feedLabelContainer {
+  display: grid;
+  grid-template-columns: repeat(auto-fill, minmax(80px, 1fr));
+  grid-gap: 5px;
+  grid-auto-flow: dense;
+  width: 100%;
+  align-items: center;
```
Please alphabetize

miles-grant-ibigroup replied:
Done in a1bc3e9

evansiroky commented on lib/style.css:
```diff
@@ -141,6 +143,114 @@ td.feed-source-info {
+  align-items: center;
+}
+.feedLabelContainer.large {
+  grid-template-columns: repeat(auto-fill, minmax(180px, 1fr));
+  grid-gap: 15px;
```
Please alphabetize

miles-grant-ibigroup replied:
Done in a1bc3e9

evansiroky commented on lib/style.css:
```diff
@@ -141,6 +143,114 @@ td.feed-source-info {
+  margin: 0 0 10px 0;
+}
+.feedSourceLabelRow {
+  display: inline-flex;
+  width: 100%;
+  align-items: baseline;
```
Please alphabetize

miles-grant-ibigroup replied:
Done in a1bc3e9

evansiroky commented on lib/style.css:
```diff
@@ -141,6 +143,114 @@ td.feed-source-info {
+  align-items: baseline;
+}
+.feedLabel {
+  display: inline-block;
+  word-break: break-word;
+
+  display: flex;
+  align-items: center;
+  justify-content: space-between;
+
+  border-radius: 5px;
+  border: 1px solid;
+
+  padding: 5px 0 8px 0;
+  height: 100%;
```
Please alphabetize

miles-grant-ibigroup replied:
Done in a1bc3e9

evansiroky commented on lib/style.css:
```diff
@@ -141,6 +143,114 @@ td.feed-source-info {
+}
+.feedLabel.small {
+  /* removing bottom padding seems to be only way to account for the icon baseline fix */
+  padding: 2px 4px 0 4px;
+  align-items: baseline;
+  justify-content: center;
```
Please alphabetize

miles-grant-ibigroup replied:
Done in a1bc3e9

evansiroky commented on lib/style.css:
```diff
@@ -141,6 +143,114 @@ td.feed-source-info {
+  margin-bottom: 2px;
+}
+.feedLabel .labelName {
+  flex: 2;
+  display: flex;
+  justify-content: center;
+  align-items: baseline;
+  padding: 5px;
```
Please alphabetize

miles-grant-ibigroup replied:
Done in a1bc3e9

evansiroky commented on lib/style.css:
```diff
@@ -141,6 +143,114 @@ td.feed-source-info {
+  padding: 5px;
+}
+.feedLabel.small .labelName {
+  padding: 0;
+  height: 100%;
+  align-items: baseline;
```
Please alphabetize

miles-grant-ibigroup replied:
Done in a1bc3e9

evansiroky commented on lib/style.css:
```diff
@@ -141,6 +143,114 @@ td.feed-source-info {
+
+/* Label Buttons */
+.feedLabel .actionButtons {
+  display: flex;
+  align-items: flex-end;
+  flex-direction: column;
```
Please alphabetize

miles-grant-ibigroup replied:
Done in a1bc3e9

evansiroky commented on lib/style.css:
```diff
@@ -141,6 +143,114 @@ td.feed-source-info {
+.feedLabel .actionButtons button {
+  background: none;
+  border: none;
+  text-decoration: underline;
+  opacity: 0.7;
```
Please alphabetize

evansiroky commented on lib/style.css:
```diff
@@ -141,6 +143,114 @@ td.feed-source-info {
+.labelActionButton {
+  background: none;
+  border: none;
+  margin: 0 0 10px 0;
+  padding: 0;
+  color: #337ab7;
+  text-decoration: none;
```
Please alphabetize

evansiroky commented on lib/manager/actions/visibilityFilter.js:
```diff
@@ -6,6 +6,14 @@ export const setVisibilitySearchText = createAction(
 )
+export const setVisibilityLabel = createAction(
+  'SET_PROJECT_VISIBILITY_LABEL',
+  (payload: Array<any>) => payload
```
Please use a more specific type than `any`.

miles-grant-ibigroup replied:
Took me a while to figure out what broke when I fixed this but got it all working in a1bc3e9

evansiroky commented on lib/manager/components/LabelPanel.js:
```diff
@@ -0,0 +1,63 @@
+    // make sure there is a valid
+    // project we can add labels tobefore continuing
+    if (!project || !project.labels || !project.id) {
+      return <div />
```
It might be clearer if null was returned instead of an empty div.

miles-grant-ibigroup replied:
I've changed this in a1bc3e9, but doesn't this sometimes cause white screens?

evansiroky commented on lib/manager/components/FeedSourceTableRow.js:
```diff
@@ -124,46 +119,69 @@ function relativeDuration (time: any) {
 }
 
+// This method is extracted variable generation which had to be split
+// out of the render function to appease the eslint complexity rule
```
+1

evansiroky commented on lib/manager/components/GeneralSettings.js:
```diff
@@ -199,6 +201,9 @@ export default class GeneralSettings extends Component<Props, State> {
             </ListGroupItem>
           </ListGroup>
         </Panel>
+        <Panel header={<h3>Labels</h3>}>
+          <LabelAssigner feedSource={feedSource} project={project} />
+        </Panel>
```
At first I was expecting this to be available within a separate tab pane within the ProjectSettings component. I think it makes more sense to add this in it's own tab pane that can be reached via clicking on a link container. Something like this:

<img width="966" alt="Screen Shot 2021-07-23 at 9 56 32 PM" src="https://user-images.githubusercontent.com/3112493/126857842-a044dc3b-e247-499a-a521-868c6bfe600c.png">

miles-grant-ibigroup replied:
Was basing the pr on [the mockups](https://www.sketch.com/s/8a229376-9d05-40ad-b1e6-04b4433b965f/a/pa885bp) but I do prefer what you have here. @landonreed what do you think?

landonreed replied:
Given that we're showing the label manager on the dashboard/ProjectViewer, I don't think it's also needed on the Settings tab. We really only need to show it in one place (and I think that place should be next to the list of feed sources).

miles-grant-ibigroup replied:
Sounds good. Will remove it!

miles-grant-ibigroup replied:
Do you think it should be removed from the feed source settings as well?

landonreed replied:
Oh, didn't even notice it was there. Yea, I think we should remove that panel also.

evansiroky commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,224 @@
+        <ListGroup fill>
+          <ListGroupItem>
+            <FormGroup
+              data-test-id='label-name'
```
It isn't necessary to always insert a data-test-id. These are mainly used with datatools's end-to-end tests when a specific selector is needed to reference certain elements. So since there isn't any additional e2e tests made for these labels, I wouldn't worry about adding these.

miles-grant-ibigroup replied:
Ah sounds good thank you!

evansiroky commented on lib/manager/components/ProjectFeedListToolbar.js:
```diff
@@ -288,6 +293,57 @@ export default class ProjectFeedListToolbar extends PureComponent<Props> {
+      index >= 0 ? labels.splice(index, 1) : labels.push(labelId)
+      setVisibilityLabel(labels)
+    }
+    function handleSelectLabelMode (e) {
+      const { value: newMode } = e.target
+      setVisibilityLabelMode(newMode)
+    }
```
This could be a class method handler.

miles-grant-ibigroup replied:
Extracted in a1bc3e9

evansiroky commented on lib/manager/components/ProjectFeedListToolbar.js:
```diff
@@ -288,6 +293,57 @@ export default class ProjectFeedListToolbar extends PureComponent<Props> {
+    const labels = this.props.filter.labels || []
+    const { labelsFilterMode } = filter
+
+    function handleClick (labelId: string) {
+      const index = labels.indexOf(labelId)
+      index >= 0 ? labels.splice(index, 1) : labels.push(labelId)
+      setVisibilityLabel(labels)
+    }
```
This one can also be a class method handler. Let's make them this way to be consistent and not be creating a new function on every render.

miles-grant-ibigroup replied:
Agreed. Extracted in a1bc3e9

evansiroky commented on lib/manager/components/ProjectSettingsForm.js:
```diff
@@ -359,6 +362,7 @@ export default class ProjectSettingsForm extends Component<Props, State> {
             </ListGroupItem>
           </ListGroup>
         </Panel>
+        <LabelPanel project={project} large user={user} />
```
Same comment as https://github.com/ibi-group/datatools-ui/pull/693#discussion_r675948735.

evansiroky commented on lib/manager/components/ProjectViewer.js:
```diff
@@ -175,72 +176,81 @@ export default class ProjectViewer extends Component<Props, State> {
                 </ButtonToolbar>
               </h3>
-              <ul className='list-unstyled list-inline small' style={{marginBottom: '0px'}}>
-                <li><Icon type='cloud-download' /> {project.autoFetchFeeds ? `${project.autoFetchHour}:${project.autoFetchMinute < 10 ? '0' + project.autoFetchMinute : project.autoFetchMinute}` : 'Auto fetch disabled'}</li>
```
lol, long line alert! Thanks for fixing

evansiroky commented on lib/manager/components/ProjectViewer.js:
```diff
@@ -175,72 +176,81 @@ export default class ProjectViewer extends Component<Props, State> {
+                    feedSources={project.feedSources || []}
+                  />
+                  <ExplanatoryPanel project={project} />
+                  <LabelPanel project={project} user={user} />
```
I didn't even notice this at first, but I like that it's there now that I see how it works.

miles-grant-ibigroup replied:
Yeah the formatting nearly swallowed this one up

evansiroky commented on lib/manager/components/ProjectViewer.js:
```diff
@@ -295,3 +307,18 @@ class ProjectSummaryPanel extends Component<{feedSources: Array<Feed>, project:
 }
+
+const ExplanatoryPanel = ({ project }) => {
+  // If user has more than 3 labels, hide the feed source instruction
+  if (project.labels.length <= 3) {
```
Probably a good UX decision. 👍

evansiroky commented on lib/manager/util/index.js:
```diff
@@ -397,3 +421,10 @@ export function projectHasAtLeastOneFeedWithAPublishedVersion (project: ?Project
   return val ? undefined : 'error'
 }
+
+export function stringifyLabels (labels: Array<any>): Array<any> {
+  // Convert feed source labels to id strings wanted by datatools-server,
+  // but only if we have a label object
+  if (!labels) return []
+  return labels.map(label => label.id ? label.id : label)
+}
```
Please remove if refactoring data model as recommended in https://github.com/ibi-group/datatools-server/pull/403#pullrequestreview-714190079.

evansiroky reviewed (dismissed):
See comments. Pretty good so far!

miles-grant-ibigroup commented:
Thanks! Have left a few comments open until we decide the best way forward with the label data structure. And one UI question for @landonreed. Adding a new sidebar item could also be a good opportunity to refactor away that giant ternary operator that controls which view is active.

landonreed commented on lib/common/components/FeedLabel.js:
```diff
@@ -0,0 +1,121 @@
+            />
+            <LabelEditorModal ref='editModal' label={this.props.label} />
+
+            <button onClick={this._onClickEdit}>
+              <Icon type='pencil' />
+            </button>
+            <button onClick={this._onClickDelete}>
+              <Icon type='trash' />
+            </button>
```
I don't think these crud buttons should be rendered within the label. I'd prefer that we show them to the right of the label, so that we're also rendering the labels in a uniform manner. As part of that uniform render, the overflow should be hidden with an ellipsis. I can provide some styling tweaks in a piggyback PR.

miles-grant-ibigroup replied:
Have adjusted the CSS a little bit! Let me know what you think. I think I hit all the things on your list.

landonreed commented on lib/manager/actions/labels.js:
```diff
@@ -0,0 +1,56 @@
+      .then(async (res) => {
+        const createdLabel = await res.json()
+        dispatch(fetchProject(createdLabel.projectId))
+        dispatch(fetchProjectFeeds(createdLabel.projectId))
```
Each of these CRUD actions may need to be structured differently so that deployments are also fetched. See `updateProject`:
https://github.com/ibi-group/datatools-ui/blob/4df5f69a195d0c1828a391e6dbba4befa82c5b24/lib/manager/actions/projects.js#L150-L151

miles-grant-ibigroup replied:
Thank you! Added.

binh-dam-ibigroup commented on lib/manager/components/LabelEditorModal.js:
```diff
@@ -0,0 +1,62 @@
+    const label = this.props.label || {}
+    const { projectId } = this.props
+    return (
+      <Modal show={this.state.showModal} onHide={this.close}>
```
Modal should include the `keyboard` prop (see https://react-bootstrap-v3.netlify.app/components/modal/#modals-props-modal).

miles-grant-ibigroup replied:
Didn't know about this thank you! Added.

binh-dam-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,236 @@
+                  name={'name'}
+                  onChange={this._onFormChange}
+                />
+                <FormControl.Feedback />
```
For the validation state to render, you may need to wrap the panel body with a react-bootstrap `<Form>`, and the Save button be of `type='submit'`.

If you are moving the Save button to the `<Modal>`'s footer, you'll need to attach the button using the form's `id` prop.

miles-grant-ibigroup replied:
Have moved the buttons to a `Footer`, and fixed the validation with a function modification. However, I haven't moved the buttons to the Modal 
1) to preserve the independence of the LabelEditor, and 
2) to avoid too much props/state passing (with the validation)

binh-dam-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,236 @@
+          labelIsNew: true,
+          // Default to a nice orange
+          // we can safely set a partial object here
+          // $FlowFixMe
```
Or you can pass the default validation state.

miles-grant-ibigroup replied:
Much cleaner thank you! Refactored.

binh-dam-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,236 @@
+    const oldLabel = this.props.label
+
+    const changes: any = {}
+    Object.keys(newLabel).map(k => {
```
`.forEach`

miles-grant-ibigroup replied:
My bad for copying from elsewhere in the code. Refactored!

binh-dam-ibigroup reviewed (changes requested):
This is cool stuff, I think we agree it's about the polishing. I am requesting changes because UI validation does not seem to work on my side (see comments so far). More comments to come.

binh-dam-ibigroup commented on lib/manager/components/LabelAssigner.js:
```diff
@@ -0,0 +1,67 @@
+ updateFeedSource: Function
+}
+
+class LabelAssigner extends React.Component<Props> {
```
Add JSDoc

miles-grant-ibigroup replied:
Thank you for the reminder. Done!

binh-dam-ibigroup commented on lib/manager/components/FeedSourceTableRow.js:
```diff
@@ -124,46 +119,69 @@ function relativeDuration (time: any) {
   )
 }
 
+// This method is extracted variable generation which had to be split
```
Sentence is incomplete. Also use `/**`-style.

miles-grant-ibigroup replied:
"Refactored"!

binh-dam-ibigroup commented on lib/manager/components/FeedSourceTableRow.js:
```diff
@@ -124,46 +119,69 @@ function relativeDuration (time: any) {
 
+// This method is extracted variable generation which had to be split
+// out of the render function to appease the eslint complexity rule
+function generateFeedSourceLabelRenderVars (feedSource, comparisonValidationSummary) {
```
Should types be added to declaration?

miles-grant-ibigroup replied:
Absolutely they should. Not sure why flow didn't catch that. Added!

landonreed commented on lib/manager/components/ProjectFeedListToolbar.js:
```diff
@@ -288,6 +307,46 @@ export default class ProjectFeedListToolbar extends PureComponent<Props> {
+            <option value='any'>any</option>
+            <option value='only'>only</option>
+          </select>{' '}
+          feeds with the labels
```
As a reader, this message doesn't make a ton of sense to me. I would change to something like `Show feeds containing [any|all] of the selected labels:`. 

Thoughts?

miles-grant-ibigroup replied:
I personally prefer any/only, but I support the wisdom of the masses. Changed!

landonreed commented on lib/manager/components/ProjectFeedListToolbar.js:
```diff
@@ -288,6 +307,46 @@ export default class ProjectFeedListToolbar extends PureComponent<Props> {
+        </h5>
+        <div className='feedLabelContainer'>
+          {project.labels.map((label) => (
+            <FeedLabel
```
As mentioned elsewhere, I would prefer the checkbox to be separated out from the label (to the left) so that we have a more consistent visual language with respect to the labels.

binh-dam-ibigroup replied:
@landonreed is your above comment regarding text flow consistent with your other comment https://github.com/ibi-group/datatools-ui/pull/693#discussion_r677445990?

landonreed replied:
I believe so, feel free to raise any specific inconsistencies, @binh-dam-ibigroup.

miles-grant-ibigroup replied:
Have moved the checkboxes out, and it feels consistent with the changes I made to the edit/delete buttons. I've changed the labels to be as you say, but there's (to my eyes) some awkwardness now with the button placement. But that may just be me. Let me know what you think! @landonreed @binh-dam-ibigroup

binh-dam-ibigroup commented on lib/manager/components/FeedSourceTableRow.js:
```diff
@@ -241,31 +266,56 @@ class FeedInfo extends PureComponent<{ feedSource: Feed }> {
+                project={project}
+              />
+              {userCanManageFeed && project.labels.length > 0 && (
+                <button
+                  className='labelActionButton small'
+                  style={{ marginRight: '5px' }}
+                  onClick={() => this.refs.labelAssignerModal.open()}
+                >
+                  Edit
+                </button>)}
+              <div className='feedLabelContainer'>
+                {feedSource.labels.map((label) => (
+                  <FeedLabel small key={label.id} label={label} />
+                ))}
```
We may have to discuss the layout and label text size in here. Do we prefer labels so they flow like text? (as opposed to vertically stacked?)

miles-grant-ibigroup replied:
I personally prefer the flowing labels, mostly because they save space. However, with the new `grid` class I added, it's easy to switch! Depends on what we want @binh-dam-ibigroup @landonreed

binh-dam-ibigroup commented on lib/manager/components/FeedSourceTableRow.js:
```diff
@@ -338,13 +388,9 @@ class FeedActionsDropdown extends PureComponent<Props> {
-      user
-    } = this.props
-
+  // This method is extracted variable generation which had to be split
```
Sentence is incomplete. Also use `/**`-style.

miles-grant-ibigroup replied:
Fixed!

binh-dam-ibigroup reviewed (commented):
Some more comments. The `LabelAssigner` component should allow checking multiple labels.

landonreed commented on lib/manager/components/ProjectFeedListToolbar.js:
```diff
@@ -368,12 +429,29 @@ export default class ProjectFeedListToolbar extends PureComponent<Props> {
+                <DropdownButton
+                  id='project-feedsource-label-filter-button'
+                  // to match the buttons with badges
+                  style={{ paddingBottom: 7 }}
```
This doesn't quite align for me. Perhaps also it would be 

![image](https://user-images.githubusercontent.com/2370911/127191809-1eeef82a-652b-47c9-995a-febc6dc889c6.png)

miles-grant-ibigroup replied:
Looks to be a text rendering issue (since it aligns on my display). Have changed the unit to `pt` in hopes that it will work a little better.

Can you let me know if this fixes it for you @landonreed?

landonreed commented on lib/manager/components/ProjectFeedListToolbar.js:
```diff
@@ -368,12 +429,29 @@ export default class ProjectFeedListToolbar extends PureComponent<Props> {
+                  title={
+                    <span
+                      style={{
+                        fontWeight: activeFilterLabelCount > 0 ? 'bold' : ''
```
This bold text is very hard to distinguish for me. I don't have an alternative recommendation at the moment.

miles-grant-ibigroup replied:
Have replaced `bold` with `900`, which should help! Otherwise I was thinking either coloring it or putting a little dot next to labels or something. The only issue is that the bar is so long it is very close to breaking over to a new line, so horizontal space is at a premium

landonreed commented on lib/manager/components/ProjectSettings.js:
```diff
@@ -1,22 +1,27 @@
+import { Row, Col, Panel, ListGroup, ListGroupItem } from 'react-bootstrap'
+import { LinkContainer } from 'react-router-bootstrap'
+
+import { deleteProject, updateProject } from '../actions/projects'
```
It's not done here, but we generally import actions into components using the `import * as projectActions from '../actions/projects'` syntax and reference actions for flow types like:

```
updateProject: typeof projectActions.updateProject
```

miles-grant-ibigroup replied:
Interesting! I prefer the style you've mentioned. Refactored!

landonreed commented on lib/manager/components/ProjectFeedListToolbar.js:
```diff
@@ -231,6 +236,20 @@ export default class ProjectFeedListToolbar extends PureComponent<Props> {
+  _onLabelClick = (labelId: string) => {
+    // This variable must be set directly from this.props both
+    // to allow the override and for flow
+    const labels = this.props.filter.labels || []
```
If we just set the default reducer state for filter.labels to `[]`, we shouldn't have to introduce this conditional here.

miles-grant-ibigroup replied:
Much cleaner thank you!

landonreed commented on lib/manager/reducers/projects.js:
```diff
@@ -23,7 +23,9 @@ export const defaultState = {
-    searchText: null
+    searchText: null,
+    // Used by the label filtering dropdown
+    labelsFilterMode: 'any'
```
Per https://github.com/ibi-group/datatools-ui/pull/693/files#r677627210, add default state for `filters: []`

miles-grant-ibigroup replied:
Feels much better. Thanks!

landonreed reviewed (changes requested):
This is looking good! Main comments are about having a uniform way to render the labels. Having a bunch of differently sized rectangles and injecting checks/buttons into them feels weird to me.

miles-grant-ibigroup commented on lib/manager/components/LabelAssigner.js:
```diff
@@ -0,0 +1,67 @@
+    const fsLabelIds = feedSource.labels.map((label) => label.id)
+
+    return (
+      <div className='feedLabelContainer large'>
```
It could be, but as far as I can tell `<React.Select>` doesn't let you see all labels/options at once! Is that something we're willing to give up? My upcoming commit moves the checkboxes outside the label, and I'm hoping that meets our needs.

binh-dam-ibigroup commented on lib/manager/components/ProjectViewer.js:
```diff
@@ -295,3 +307,18 @@ class ProjectSummaryPanel extends Component<{feedSources: Array<Feed>, project:
+
+const ExplanatoryPanel = ({ project }) => {
+  // If user has more than 3 labels, hide the feed source instruction
+  if (project.labels.length <= 3) {
```
(non-blocking) I'd be in favor of leaving the condition outside of this component.

miles-grant-ibigroup replied:
I originally had this, but this increases the complexity of the main render method beyond what eslint allows

binh-dam-ibigroup commented on lib/manager/util/index.js:
```diff
@@ -20,6 +18,8 @@ import type {
   ProjectsState
 } from '../../types/reducers'
 
+import {getVersionValidationSummaryByFilterStrategy} from './version'
+
```
Undo this change.

miles-grant-ibigroup replied:
This change is made automatically during the eslint pre-commit hook. Is my eslint config broken/outdated? It's the one inside the mastarm folder inside `node_modules`

binh-dam-ibigroup commented on lib/manager/util/index.js:
```diff
@@ -397,3 +421,10 @@ export function projectHasAtLeastOneFeedWithAPublishedVersion (project: ?Project
 }
+
+export function stringifyLabels (labels: Array<any>): Array<any> {
+  // Convert feed source labels to id strings wanted by datatools-server,
+  // but only if we have a label object
```
Convert to JSDoc.

miles-grant-ibigroup replied:
Done!

binh-dam-ibigroup commented on lib/types/reducers.js:
```diff
@@ -1,9 +1,10 @@
 
 import UserPermissions from '../common/user/UserPermissions'
 import UserSubscriptions from '../common/user/UserSubscriptions'
+import type { PermissionType } from '../common/user/UserPermissions'
```
Maybe insert blank line before the type imports.

miles-grant-ibigroup replied:
Same note as https://github.com/ibi-group/datatools-ui/pull/693#discussion_r677893400. Reverting it is undone during the pre-commit hook.

binh-dam-ibigroup commented on lib/manager/actions/labels.js:
```diff
@@ -0,0 +1,56 @@
+  // TODO: is there a cleaner/more dynamic way to do this? Properties can't be guaranteed to
+  // include all the keys we need, so can't use that
+
+  // $FlowFixMe
```
Maybe instead of deleting fields from the argument, do this?
```js
const { organizationId, user, ...cleanedLabel } = label
```

miles-grant-ibigroup replied:
This is much better thank you! Much appreciated

binh-dam-ibigroup commented on lib/manager/components/LabelPanel.js:
```diff
@@ -0,0 +1,63 @@
+import LabelEditorModal from './LabelEditorModal'
+
+export type Props = {
+  large?: boolean,
```
What does `large` mean? Wide or tall?

miles-grant-ibigroup replied:
It means both! Although less now after feedback from Landon

binh-dam-ibigroup commented on lib/manager/components/ProjectViewer.js:
```diff
@@ -175,72 +176,81 @@ export default class ProjectViewer extends Component<Props, State> {
+                    feedSources={project.feedSources || []}
+                  />
+                  <ExplanatoryPanel project={project} />
+                  <LabelPanel project={project} user={user} />
```
Labels in the ProjectViewer should not be this tall, even with the label Edit and Delete buttons visible.

miles-grant-ibigroup replied:
New CSS shrinks them a little, but am open to more feedback on the matter.

binh-dam-ibigroup replied:
Make the text on these the same size as in the labels filter dropdown, or the label Edit button. At the current size, the labels feel a bit overwhelming compared to the other content.

binh-dam-ibigroup reviewed (commented):
Some more comments. Let me know about the decision on label appearance.

miles-grant-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,236 @@
+                      <ControlLabel>Visible to admins only?</ControlLabel>
+                      <Row>
+                        <Col xs={3}>
+                          <FormControl
```
Good call thanks! Have adjusted this and it looks much more coherent.

miles-grant-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,236 @@
+                          />
+                        </Col>
+                        <Col xs={9}>
+                          <div>
```
I restructured this entire component, and there should be fewer divs now, beyond what's needed for all the flex box css.

miles-grant-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,236 @@
+                  </FormGroup>
+                </Col>
+                <Col xs={9}>
+                  {this.userIsAdmin(user) ? (
```
Definitely does. Good catch thank you!

miles-grant-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,236 @@
+            </ListGroupItem>
+          </ListGroup>
+        </Panel>
+        <Row>
```
See https://github.com/ibi-group/datatools-ui/pull/693#discussion_r677889581

miles-grant-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,236 @@
+  }
+}
+
+class LabelEditor extends React.Component<Props, State> {
```
Added

miles-grant-ibigroup commented on lib/common/components/FeedLabel.js:
```diff
@@ -0,0 +1,121 @@
+}
+export type State = {}
+
+// Generate lightened/darkened versions of a color for use in text
```
Thank you for the reminder. Done!

miles-grant-ibigroup commented on lib/common/components/FeedLabel.js:
```diff
@@ -0,0 +1,121 @@
+  return complementary.toHexString()
+}
+
+class FeedLabel extends React.Component<Props, State> {
```
Thank you for the reminder. Done!

miles-grant-ibigroup commented on lib/common/components/FeedLabel.js:
```diff
@@ -0,0 +1,121 @@
+      user.permissions.isProjectAdmin(label.projectId)
+
+    return (
+      <div
```
I tried this, and while it would probably help, the Bootstrap `label` stylings interfere too strongly with the custom styling, and I don't want to complicate the CSS any more.

miles-grant-ibigroup replied:
We can do it anyway though if you think it is a good idea

binh-dam-ibigroup replied:
I'm talking about the plain HTML `<label>` that goes with `<input>` controls.

miles-grant-ibigroup replied:
I think there is global bootstrap css which causes even those to be styled. It doesn't look to be more than padding and font weight, but I am weary of adding even more styling. Do you think it's worth it?

miles-grant-ibigroup commented on lib/manager/components/LabelAssigner.js:
```diff
@@ -0,0 +1,67 @@
+      updatedLabels.splice(labelIndex, 1)
+    }
+
+    this.props.updateFeedSource(feedSource, { labels: updatedLabels })
```
Have changed this now to only update on component unmount. Ideally, the labels would update in the background as you click, and there is a risk of reloading and losing changes, but I think this is a good start?

miles-grant-ibigroup commented on lib/common/components/FeedLabel.js:
```diff
@@ -0,0 +1,121 @@
+          {checked !== undefined && (
+            <div className='actionButtons'>
+              <input type='checkbox' checked={checked} />
+            </div>
```
With flex box they keep coming and coming...

One of the reasons I wanted to go with grid to start with, but the design we're shooting for doesn't work so well with grid :(

miles-grant-ibigroup commented on lib/manager/components/LabelPanel.js:
```diff
@@ -0,0 +1,63 @@
+
+    return (
+      <Panel header={<h3>Labels</h3>}>
+        <div className={`feedLabelContainer ${large ? 'large' : ''}`}>
```
Unfortunately, this breaks the grid. Needs to be it's own div inside the "white space" of the panel

miles-grant-ibigroup commented on lib/manager/components/ProjectFeedListToolbar.js:
```diff
@@ -231,6 +236,20 @@ export default class ProjectFeedListToolbar extends PureComponent<Props> {
+    index >= 0 ? labels.splice(index, 1) : labels.push(labelId)
+    this.props.setVisibilityLabel(labels)
+  }
+  _onLabelModeSelectorClick = (e: { target: {value: string} }) => {
```
This is what I was looking for thank you!

miles-grant-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,236 @@
+  }
+
+  // compares the dirty label from the old one to determine
+  // which keys changed
```
Done!

miles-grant-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,236 @@
+    })
+    return changes
+  }
+  _settingsAreUnedited = () => Object.keys(this._getChanges()).length === 0
```
Good catch thank you. Added!

miles-grant-ibigroup commented on lib/manager/components/LabelEditorModal.js:
```diff
@@ -0,0 +1,62 @@
+  showModal: boolean,
+};
+
+export default class LabelEditorModal extends React.Component<
```
Good catch thank you! Added.

miles-grant-ibigroup commented on lib/manager/components/LabelEditorModal.js:
```diff
@@ -0,0 +1,62 @@
+type Props = {
+  label?: Label,
+  projectId?: String,
+};
```
Eslint doesn't seem to care about them... Have removed them but will look into if it is my eslintrc that is broken or if something else is wrong

miles-grant-ibigroup commented on lib/manager/components/LabelEditorModal.js:
```diff
@@ -0,0 +1,62 @@
+
+  ok = () => {
+    this.close()
+  };
```
They are used via a ref. I agree it is not the best way to do it, and this should be fixed with the move to React 16+

miles-grant-ibigroup commented on lib/manager/components/LabelEditorModal.js:
```diff
@@ -0,0 +1,62 @@
+  render () {
+    const { Body, Header, Title } = Modal
+    const label = this.props.label || {}
+    const { projectId } = this.props
```
As far as I can tell, you can't do the conditional when using that syntax? Or am I mistaken

miles-grant-ibigroup commented on lib/manager/components/ProjectViewer.js:
```diff
@@ -175,72 +176,81 @@ export default class ProjectViewer extends Component<Props, State> {
                 <ButtonToolbar className={`pull-right`}>
-                  {getConfigProperty('application.notifications_enabled')
-                    ? <WatchButton
+                  {getConfigProperty('application.notifications_enabled') ? (
```
I prefer prettier's syntax but have removed to avoid noise

miles-grant-ibigroup commented on lib/manager/components/ProjectViewer.js:
```diff
@@ -175,72 +176,81 @@ export default class ProjectViewer extends Component<Props, State> {
-                      />
-                    )
-                  }
+                  {this.state.createMode ? (
```
Have removed to avoid noise

miles-grant-ibigroup commented:
Thanks all for the feedback! Have addressed most comments, and responded to the rest. 

In terms of the label rendering, I personally like the flowing style we have on the individual feed labels. I am less a fan of the awkward edit/delete buttons bouncing around in the label panel. Not sure what the best fix is. I tried forcing them all to the right, but the changing white space is awkward too... Could move them to the left but that doesn't flow nicely (too many icons on one side when the label is admin only).

binh-dam-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,231 @@
+                <HelpBlock>Required.</HelpBlock>
+              </FormGroup>
+              <FormGroup>
+                <ControlLabel>Description</ControlLabel>
```
When editing an existing label, changing fields other than Description should also enable the Save button.

miles-grant-ibigroup replied:
This was a good catch! I've refactored the checkbox validation to resolve this

binh-dam-ibigroup commented on lib/common/components/FeedLabel.js:
```diff
@@ -0,0 +1,132 @@
+      user.permissions.isProjectAdmin(label.projectId)
+
+    return (
+      <div className='feedLabelWrapper'
```
See markup suggestions in #697 (padding and some alignment is no perfect in there)

miles-grant-ibigroup replied:
Does react allow you to change the top level component type based on a variable? Also there is this issue with doing that: https://github.com/ibi-group/datatools-ui/pull/693#discussion_r678443261

binh-dam-ibigroup commented on lib/manager/actions/labels.js:
```diff
@@ -0,0 +1,57 @@
+}
+
+/**
+ * Permanently delete single label */
```
Move `*/` to new line.

miles-grant-ibigroup replied:
Moved!

binh-dam-ibigroup commented on lib/manager/components/FeedSourceTableRow.js:
```diff
@@ -124,46 +120,72 @@ function relativeDuration (time: any) {
   )
 }
 
+//
```
Remove this line.

miles-grant-ibigroup replied:
Removed!

binh-dam-ibigroup commented on lib/manager/components/FeedSourceTableRow.js:
```diff
@@ -124,46 +120,72 @@ function relativeDuration (time: any) {
 
+//
+/**
+ * This method is extracted variable generation which had to be split out
```
Sentence does not make sense to me.

miles-grant-ibigroup replied:
Have reworded it and I hope it makes more sense now!

binh-dam-ibigroup commented on lib/manager/components/LabelEditorModal.js:
```diff
@@ -0,0 +1,69 @@
+    })
+  };
+
+  // Used in the ref
```
I can't recall a recent extensive use of refs like this for this repo... Is there an alternate approach?

miles-grant-ibigroup replied:
I based the Modal files on https://github.com/ibi-group/datatools-ui/blob/dev/lib/common/components/InfoModal.js

My original approach used React hooks, which aren't available in the version of React we're using!

binh-dam-ibigroup reviewed (changes requested):
Please consider suggestions from #697 and advise regarding the use of refs. See also other formatting comments.

miles-grant-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,231 @@
+                                          If checked, this label will only be visible to
+                                          project and application admins. Users with
+                                          feed-specific priveleges will not see this label
+                                          applied to any feed sources.
```
Have removed thanks! And have adopted the `Modal.Footer` component. Looks good thank you!

miles-grant-ibigroup commented on lib/manager/components/LabelAssigner.js:
```diff
@@ -0,0 +1,96 @@
+      <div className='feedLabelContainer large'>
+        {projectLabels.length === 0 && (
+          <div className='noLabelsMessage'>
+        There are no labels in this project.
```
Have indented.

miles-grant-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,231 @@
+                <Col xs={3}>
+                  <FormGroup>
+                    <ControlLabel>Color</ControlLabel>
+                    <FormControl
```
This was an issue on the server side. Has been resolved in https://github.com/ibi-group/datatools-server/pull/403/commits/cdb706180102cee4a377ec0bae084af5758a6edb

miles-grant-ibigroup commented on lib/manager/components/LabelEditorModal.js:
```diff
@@ -0,0 +1,69 @@
+
+  render () {
+    const { Body, Header, Title } = Modal
+    const label = this.props.label || {}
```
This was an issue on the server side. Has been resolved in https://github.com/ibi-group/datatools-server/pull/403/commits/cdb706180102cee4a377ec0bae084af5758a6edb

miles-grant-ibigroup commented on lib/common/components/FeedLabel.js:
```diff
@@ -0,0 +1,132 @@
+        tabIndex={0}
+      >
+        {checked !== undefined && (
+          <div className='actionButtons'>
```
I will play with your PR a little bit to see if I can get something sensible, but from what I can gather making a `<span>` `display:flex` doesn't work properly, and the result is the weird padding issues you mention in a later comment.

miles-grant-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,231 @@
+    )
+  }
+
+  /* Compares the dirty label from the old one to determine
```
Fixed!

landonreed commented:
I have just added some changes in #698.

miles-grant-ibigroup commented:
To aid the transition to the new PR policy, I'll let everyone know what's going on. In the future, since we'll only do 1 reviewer at a time, this won't be necessary but want to keep everyone on the same page!

Going to first assign @binh-dam-ibigroup. When all of Binh's feedback is addressed, I'm going to assign @landonreed.

binh-dam-ibigroup commented on lib/common/components/FeedLabel.js:
```diff
@@ -48,52 +48,78 @@ class FeedLabel extends React.Component<Props, State> {
-  // to be true/false
-      checked !== undefined ? 'clickable' : ''
-      }`}
+        <label
```
Why not, although for semantics, when there is no checkbox, the label refers to an element that does not exist (that's why I had it in a particular way in #697).

miles-grant-ibigroup replied:
Ah this is a good point. Could try to make the for field optional? Or could make the syntactic argument that it's technically still a label for a certain checkbox it's just that that checkbox doesn't exist haha

binh-dam-ibigroup replied:
In #697 I had two separate renderings one when a checkbox was displayed and one otherwise.

miles-grant-ibigroup replied:
Tried to do this the cool way and have optional props via spread, but our accessibility eslint rule required a for for a label. 

@landonreed do you think it's worth doing a second render option?

landonreed replied:
Nope, I'm fine with it as is.

binh-dam-ibigroup commented on lib/common/components/FeedLabel.js:
```diff
@@ -48,52 +48,78 @@ class FeedLabel extends React.Component<Props, State> {
-        tabIndex={0}
-      >
-        {checked !== undefined && (
+      <div className={this._getWrapperClasses()}>
```
All `<div>`s in here should be `<span>`s (see https://dev.to/kenbellows/stop-using-so-many-divs-an-intro-to-semantic-html-3i9i) and I do feel strongly about that.

binh-dam-ibigroup replied:
(For features of `<div>` that `<span>` does not provide, use CSS.)

miles-grant-ibigroup replied:
Good read thank you!

binh-dam-ibigroup commented on lib/manager/components/LabelPanel.js:
```diff
@@ -36,15 +37,12 @@ export default class LabelPanel extends Component<Props> {
 
     return (
-      <Panel header={<h3>Labels</h3>}>
+      <Panel header={<h3><Icon type={'tag'} />{' '}Labels</h3>}>
```
Good idea!

miles-grant-ibigroup replied:
Thank Landon not me :)

binh-dam-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,234 @@
+                <ControlLabel>Description</ControlLabel>
+                <FormControl
+                  value={newLabel.description}
+                  name={'description'}
```
When creating a new label, if you change the description the modal will "let" you save the label without entering a name (that's caught server-side).

![Screen Recording 2021-07-30 at 12 25 05 PM](https://user-images.githubusercontent.com/56846598/127683383-de0b1e29-acf8-478d-9753-a546482afcf8.gif)

miles-grant-ibigroup replied:
Really good catch thanks! Have refactored the validation a bunch

binh-dam-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,234 @@
+        <Modal.Footer>
+          {/* Cancel button */}
+          <Button onClick={this.props.onDone} style={{ marginRight: 10 }}>
+              Cancel
```
Fix indent

miles-grant-ibigroup replied:
This one was done by prettier but I've put it all one one line for now

binh-dam-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -0,0 +1,234 @@
+            type='submit'
+            onClick={this._onSaveSettings}
+          >
+              Save
```
Fix indent

binh-dam-ibigroup commented on lib/manager/components/FeedSourceTableRow.js:
```diff
@@ -338,13 +391,15 @@ class FeedActionsDropdown extends PureComponent<Props> {
-    } = this.props
-
+  /**
+   * This method is extracted variable generation which had to be split
```
There was another instance of this sentence that doesn't make sense that I think you fixed.

binh-dam-ibigroup reviewed (changes requested):
In `FeedLabel`, convert all `<div>` to `<span>`, and `LabelEditor` lets you add a label without entering a name.

binh-dam-ibigroup commented on lib/manager/components/FeedSourceTableRow.js:
```diff
@@ -392,8 +392,8 @@ class FeedActionsDropdown extends PureComponent<Props> {
   /**
-   * This method is extracted variable generation which had to be split
-   * out of the render function to appease the eslint complexity rule
+   * This method is variable generation which used to live in the render method,
```
Sorry I should have mentioned its the term "variable generation" that didn't make sense to me.

binh-dam-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -114,8 +112,12 @@ class LabelEditor extends React.Component<Props, State> {
+    const {validation, newLabel} = this.state
+
+    return Object.keys(validation).every(k => validation[k]) &&
+    // Resolve edge case when creating a new label
+    newLabel.name &&
+    newLabel.name !== ''
```
This might need to be indented.

miles-grant-ibigroup replied:
Have run it through Prettier!

binh-dam-ibigroup commented on lib/manager/components/LabelEditor.js:
```diff
@@ -209,7 +204,7 @@ class LabelEditor extends React.Component<Props, State> {
             onClick={this._onSaveSettings}
           >
-              Save
+            Save
```
Is it supposed to trigger a full page refresh? @landonreed what do you think?

miles-grant-ibigroup replied:
It used to only refresh feed sources, but then at Landon's suggestion we had it refresh project deployments as well which causes the full page refresh. I am not sure why we fetch project deployments, but I trust Landon since he's worked with this a lot longer than I have

binh-dam-ibigroup reviewed (approved):
Thanks for the changes @miles-grant-ibigroup. Just wondering about the page reload thing, but others can weigh in on that.

miles-grant-ibigroup commented:
Thanks @binh-dam-ibigroup! Have addressed all suggestions, will now assign to Landon

landonreed commented on lib/manager/components/LabelEditor.js:
```diff
@@ -72,24 +72,22 @@ class LabelEditor extends React.Component<Props, State> {
-      })
-    )
-  }
+    let valid = targetValue && targetValue.length > 0
```
This is 11th hour, but something I realized was that we should probably limit the label length to something like 25 characters. Probably OK just to handle this with `<input maxlength="25" ... />`.

landonreed replied:
Handled in #700.

landonreed reviewed (approved):
This all looks really good! Excellent work, @miles-grant-ibigroup! Approval contingent on merging #700

miles-grant-ibigroup commented:
Thanks Landon! Have merged and made a related server commit https://github.com/ibi-group/datatools-server/pull/403/commits/c4ba8a1f75732645a864edd3b08e745afa1ce1de

Not sure how we wound up with 3 reviewers, but for something this big I guess it can't hurt? Assigning to Evan

evansiroky commented:
Nah, I trust the reviews done, so I won't take a further look. I'm assigning back to @miles-grant-ibigroup so he can coordinate with the release manager... who is... ?

miles-grant-ibigroup commented:
@evansiroky I think it's still blocked waiting for you to click approve. I will keep this assigned to myself for now until https://github.com/ibi-group/datatools-server/pull/403 is ready for merge

landonreed commented:
I thought dismissing Evan's review would work, but it still shows blocked. Re-assigning Evan to please click the green button.
